### PR TITLE
ENHANCEMENT Backport of querystring work to 3.x

### PR DESCRIPTION
**Pull request by tractorcow**
Parent issue https://github.com/silverstripe/silverstripe-versioned/issues/145

### CODE REVIEW

File: dev/FunctionalTest.php [hunk 1]
```diff
@@ -379,11 +379,11 @@ public function logInAs($member) {
 	 */
 	public function useDraftSite($enabled = true) {
 		if($enabled) {
-			$this->session()->inst_set('readingMode', 'Stage.Stage');
```
Comment (tractorcow) [inline dev/FunctionalTest.php]: Doh, phpstorm refactor messed up all these strings.

**Comment by tractorcow** [comment]
So far these errors locally. Should be easy to fix.

```
Time: 7.26 minutes, Memory: 86.00MB

There were 2 errors:

1) HtmlEditorFieldTest::testGetAnchors
SS_HTTPResponse_Exception: You are not permitted to access the content of the target page.

/Users/dmooyman/Sites/ss37test/framework/forms/HtmlEditorField.php:581
/Users/dmooyman/Sites/ss37test/framework/tests/forms/HtmlEditorFieldTest.php:165
/Users/dmooyman/Sites/ss37test/vendor/phpunit/phpunit/composer/bin/phpunit:63

2) UploadFieldTest::testConfigOverwriteWarningCannotRelaceFiles
InvalidArgumentException: Invalid stage name ""

/Users/dmooyman/Sites/ss37test/framework/model/VersionedReadingMode.php:139
/Users/dmooyman/Sites/ss37test/framework/model/VersionedReadingMode.php:31
/Users/dmooyman/Sites/ss37test/framework/model/Versioned.php:237
/Users/dmooyman/Sites/ss37test/framework/core/Object.php:1068
/Users/dmooyman/Sites/ss37test/framework/model/DataQuery.php:163
/Users/dmooyman/Sites/ss37test/framework/model/DataQuery.php:50
/Users/dmooyman/Sites/ss37test/framework/model/DataList.php:55
/Users/dmooyman/Sites/ss37test/framework/model/ManyManyList.php:55
/Users/dmooyman/Sites/ss37test/framework/control/injector/InjectionCreator.php:17
/Users/dmooyman/Sites/ss37test/framework/control/injector/Injector.php:560
/Users/dmooyman/Sites/ss37test/framework/control/injector/Injector.php:869
/Users/dmooyman/Sites/ss37test/framework/control/injector/Injector.php:907
/Users/dmooyman/Sites/ss37test/framework/core/Object.php:142
/Users/dmooyman/Sites/ss37test/framework/model/DataObject.php:1798
/Users/dmooyman/Sites/ss37test/framework/core/Object.php:764
/Users/dmooyman/Sites/ss37test/cms/code/model/SiteTreeFileExtension.php:111
/Users/dmooyman/Sites/ss37test/framework/core/Object.php:1068
/Users/dmooyman/Sites/ss37test/framework/model/DataObject.php:1150
/Users/dmooyman/Sites/ss37test/framework/model/DataObject.php:1488
/Users/dmooyman/Sites/ss37test/framework/tests/forms/uploadfield/UploadFieldTest.php:793
/Users/dmooyman/Sites/ss37test/vendor/phpunit/phpunit/composer/bin/phpunit:63

--


There was 1 failure:

1) DirectorTest::testRequestFilterInDirectorTest
Failed asserting that exception of type "InvalidArgumentException" matches expected exception "SS_HTTPResponse_Exception". Message was: "Invalid stage name """.

/Users/dmooyman/Sites/ss37test/vendor/phpunit/phpunit/composer/bin/phpunit:63

FAILURES!
Tests: 1800, Assertions: 7331, Failures: 1, Errors: 2, Incomplete: 6, Skipped: 24.
```

**Comment by tractorcow** [comment]
Fixed and ready to review.

File: forms/Form.php [hunk 2]
```diff
@@ -1070,11 +1070,24 @@ public function getStrictFormMethodCheck() {
...
 			return $this->formActionPath;
-		} elseif($this->controller->hasMethod("FormObjectLink")) {
-			return $this->controller->FormObjectLink($this->name);
+		}
+
+		// Forms without parent controller have no link;
+        // E.g. Submission handled via graphql
+		if (!$this->controller) {
+			return null;
+		}
```
Comment (unclecheese) [inline forms/Form.php]: Unlike 4.0, `$controller` is a required constructor arg in 3.x. Is this only reachable with:

```php
$form = new Form($this ...);
$form->setController(null);
```
??

Also might want to remove the reference to graphql, as it has no compatibility with 3.0 and likely never will.

Comment (tractorcow) [reply forms/Form.php]: Ah right, that was a blind copy paste sorry. You're right.

**Comment by unclecheese** [comment]
This can be merged ahead of green, as the [Postgres test failure] is an open issue (https://github.com/silverstripe/silverstripe-framework/issues/8041)

**Comment by dhensby** [comment]
I'm experiencing a regression here where a dev/build causes a redirect to `/Security/login?BackURL=%2Fpage-not-found%2F%3Fstage%3DStage`

**Comment by tractorcow** [comment]
Raised that as a bug at https://github.com/silverstripe/silverstripe-framework/issues/8119

**Pull request merged**